File: generate_report.py
#!/usr/bin/env python3
import json
import os
from datetime import datetime

TODAY = "2026-03-02"
OUTPUT_DIR = "/Users/ozlemsultan/.openclaw/workspace/reddit-productivity/daily"
RAW_JSON = os.path.join(OUTPUT_DIR, "_temp_posts.json")
REPORT_FILE = os.path.join(OUTPUT_DIR, f"{TODAY}.md")
MEDIA_DIR = f"{TODAY}-media"

KEYWORDS = ["productivity", "focus", "screen time", "gamification", "phone addiction", "study", "discipline", "digital minimalism", "streak", "accountability", "motivation"]

def calculate_viral_score(post):
    score = post.get('score', 0)
    num_comments = post.get('num_comments', 0)
    upvote_ratio = post.get('upvote_ratio', 1.0)
    text = (post.get('title','') + " " + post.get('selftext','')).lower()

    # Raw points (0-10)
    raw_points = min(score, 5000) / 500

    # Engagement points (cap 3, later scaled)
    engagement_raw = (num_comments / (score + 1)) * 100 * 0.03
    engagement_points = min(engagement_raw, 3)

    # Upvote ratio points (max 10)
    ratio_points = upvote_ratio * 10

    # Relevance bonus (max 3)
    found_keywords = set()
    for kw in KEYWORDS:
        if kw in text:
            found_keywords.add(kw)
    relevance_bonus = min(len(found_keywords), 3)

    total = raw_points + engagement_points + ratio_points + relevance_bonus
    viral_score = round(total / 2.6, 1)
    return {
        'viral_score': viral_score,
        'raw_points': raw_points,
        'engagement_points': engagement_points,
        'ratio_points': ratio_points,
        'relevance_bonus': relevance_bonus
    }

# Load posts
with open(RAW_JSON, 'r') as f:
    posts = json.load(f)

# Compute scores
for post in posts:
    scores = calculate_viral_score(post)
    post.update(scores)

# Sort by viral_score descending
posts.sort(key=lambda x: x['viral_score'], reverse=True)

# Write report
with open(REPORT_FILE, 'w', encoding='utf-8') as f:
    f.write(f"# Reddit Scout Report: Focus Timer Opportunities\n")
    f.write(f"**Date:** {TODAY}\n\n")

    f.write("## Top Opportunities\n\n")

    # Top 5
    for idx, post in enumerate(posts[:5], 1):
        hours_ago = round((datetime.now().timestamp() - post['created']) / 3600, 1)
        f.write(f"### {idx}. [{post['title']}](https://www.reddit.com{post['permalink']})\n")
        f.write(f"Subreddit: r/{post['subreddit']} | Score: {post['score']} | Comments: {post['num_comments']} | Upvote ratio: {post['upvote_ratio']:.0%}\n")
        f.write(f"Posted: ~{hours_ago} hours ago\n\n")
        summary = post['selftext'].strip()
        if not summary:
            summary = "(No text content)"
        # Make 2-3 sentences summary - just use first 200 chars as per instructions
        f.write(f"**Summary:** {summary}\n\n")
        f.write(f"**Viral Score:** {post['viral_score']:.1f}/10\n")
        f.write(f"- Raw score: {post['raw_points']:.1f}/10\n")
        f.write(f"- Engagement: {post['engagement_points']:.1f}/10\n")
        f.write(f"- Upvote ratio: {post['ratio_points']:.1f}/10\n")
        f.write(f"- Relevance bonus: {post['relevance_bonus']}/3\nn")
        f.write("**Media:**\n")
        if post['media_sources']:
            for src in post['media_sources']:
                # Use alt text based on filename
                alt = f"Image from r/{post['subreddit']}"
                f.write(f"![{alt}]({src})\n")
        else:
            f.write("(No media)\n")
        f.write("\n")

    # Honorable Mentions (6-10)
    f.write("## Honorable Mentions\n\n")
    for idx, post in enumerate(posts[5:10], 6):
        summary = post['selftext'].strip()[:100]
        if not summary:
            summary = "(No text content)"
        f.write(f"### {idx}. [{post['title']}](https://www.reddit.com{post['permalink']}) (r/{post['subreddit']} | {post['score']} upvotes) – {summary}.\n")

    # Media Summary
    f.write("\n## Media Summary\n")
    f.write(f"Downloaded images ({TODAY}-media/):\n")
    # List files in media directory with sizes
    try:
        files = os.listdir(os.path.join(OUTPUT_DIR, TODAY + "-media"))
        for file in sorted(files):
            filepath = os.path.join(OUTPUT_DIR, TODAY + "-media", file)
            size_kb = round(os.path.getsize(filepath) / 1024, 1)
            alt = f"Image from r/{TODAY}"
            f.write(f"- **{file}** ({size_kb} KB)\n")
            f.write(f"  ![Description]({TODAY}-media/{file})\n")
    except Exception as e:
        f.write(f"(Error listing media: {e})\n")

    # GitHub link at the very end
    f.write("\n---\n")
    f.write(f"**View on GitHub:** https://github.com/ozlemsultan90-cmyk/reddit-scout-reports/blob/main/reports/{TODAY}.md\n")

print(f"Report written to {REPORT_FILE}")

File: generate_report.js
const fs = require('fs');
const path = require('path');
const https = require('https');

// Determine today's date in Europe/Istanbul timezone
const now = new Date();
const options = { timeZone: 'Europe/Istanbul', year: 'numeric', month: '2-digit', day: '2-digit' };
const parts = new Intl.DateTimeFormat('en-CA', options).formatToParts(now);
const year = parts.find(p => p.type === 'year').value;
const month = parts.find(p => p.type === 'month').value;
const day = parts.find(p => p.type === 'day').value;
const today = `${year}-${month}-${day}`;

const mediaDir = `/Users/ozlemsultan/.openclaw/workspace/reddit-productivity/daily/${today}-media`;
const tempFile = `/Users/ozlemsultan/.openclaw/workspace/reddit-productivity/daily/_temp_posts.json`;
const reportFile = `/Users/ozlemsultan/.openclaw/workspace/reddit-productivity/daily/${today}.md`;

const keywords = ['productivity', 'focus', 'screen time', 'gamification', 'phone addiction', 'study', 'discipline', 'digital minimalism', 'streak', 'accountability', 'motivation'];

// Load posts
const posts = JSON.parse(fs.readFileSync(tempFile, 'utf8'));

function calculateViralScore(post) {
  const score = post.score || 0;
  const numComments = post.num_comments || 0;
  const upvoteRatio = post.upvote_ratio || 0.5;

  // Raw score (0-10)
  const rawPoints = Math.min(score, 5000) / 500;

  // Engagement (cap 3 before scaling)
  const engagementRaw = (numComments / (score + 1)) * 100 * 0.03;
  const engagementPoints = Math.min(engagementRaw, 3);

  // Upvote ratio (max 10)
  const ratioPoints = upvoteRatio * 10;

  // Relevance bonus (max 3)
  const combinedText = ((post.title || '') + ' ' + (post.selftext || '')).toLowerCase();
  const distinctMatches = new Set();
  for (const kw of keywords) {
    if (combinedText.includes(kw)) {
      distinctMatches.add(kw);
    }
  }
  const relevanceBonus = Math.min(distinctMatches.size, 3);

  // Total scaled to 0-10
  const total = rawPoints + engagementPoints + ratioPoints + relevanceBonus;
  const viralScore = Math.round((total / 2.6) * 10) / 10;

  return {
    viral_score: viralScore,
    raw_points: Math.round(rawPoints * 10) / 10,
    engagement_points: Math.round(engagementPoints * 10) / 10,
    ratio_points: Math.round(ratioPoints * 10) / 10,
    relevance_bonus: relevanceBonus
  };
}

function downloadImage(url, destPath) {
  return new Promise((resolve, reject) => {
    const file = fs.createWriteStream(destPath);
    https.get(url, (response) => {
      response.pipe(file);
      file.on('finish', () => {
        file.close();
        const stats = fs.statSync(destPath);
        if (stats.size > 1024) {
          resolve(destPath);
        } else {
          fs.unlinkSync(destPath);
          reject(new Error('File too small'));
        }
      });
    }).on('error', (err) => {
      file.close();
      if (fs.existsSync(destPath)) fs.unlinkSync(destPath);
      reject(err);
    });
  });
}

async function processPosts() {
  const enrichedPosts = [];

  for (const post of posts) {
    const { viral_score, raw_points, engagement_points, ratio_points, relevance_bonus } = calculateViralScore(post);
    const mediaSources = [];

    // Check for image to download
    const postHint = post.post_hint;
    const isGallery = post.is_gallery === true;
    const url = post.url || '';

    try {
      if (postHint === 'image' && (url.endsWith('.jpg') || url.endsWith('.png') || url.endsWith('.jpeg') || url.endsWith('.gif'))) {
        const filename = path.basename(url.split('?')[0]);
        const dest = path.join(mediaDir, filename);
        if (!fs.existsSync(dest)) {
          await downloadImage(url, dest);
        }
        mediaSources.push(`${today}-media/${filename}`);
      } else if (isGallery && post.media_metadata) {
        for (const [key, item] of Object.entries(post.media_metadata)) {
          const imgUrl = item.s?.u || item.p?.replace('&amp;', '&');
          if (imgUrl) {
            const ext = imgUrl.match(/\.(jpg|jpeg|png|gif)/i)?.[1] || 'jpg';
            const filename = `${key}.${ext}`;
            const dest = path.join(mediaDir, filename);
            if (!fs.existsSync(dest)) {
              await downloadImage(imgUrl, dest);
            }
            mediaSources.push(`${today}-media/${filename}`);
          }
        }
      } else if (url.includes('i.redd.it') && (url.endsWith('.jpg') || url.endsWith('.png') || url.endsWith('.jpeg') || url.endsWith('.gif'))) {
        const filename = path.basename(url.split('?')[0]);
        const dest = path.join(mediaDir, filename);
        if (!fs.existsSync(dest)) {
          await downloadImage(url, dest);
        }
        mediaSources.push(`${today}-media/${filename}`);
      }
    } catch (e) {
      // Continue even if image download fails
      console.error(`Failed to download image for post ${post.id}:`, e.message);
    }

    // Calculate hours ago
    const created = post.created * 1000;
    const now = Date.now();
    const hoursAgo = Math.max(0, Math.floor((now - created) / (1000 * 60 * 60)));

    enrichedPosts.push({
      ...post,
      viral_score,
      raw_points,
      engagement_points,
      ratio_points,
      relevance_bonus,
      mediaSources,
      hours_ago: hoursAgo
    });
  }

  // Sort by viral score descending
  enrichedPosts.sort((a, b) => b.viral_score - a.viral_score);

  return enrichedPosts;
}

function formatReport(topPosts) {
  const lines = [];
  lines.push(`# Reddit Scout Report: Focus Timer Opportunities`);
  lines.push(`**Date:** ${today}`);
  lines.push('');
  lines.push(`## Top Opportunities`);
  lines.push('');

  for (let i = 0; i < Math.min(5, topPosts.length); i++) {
    const post = topPosts[i];
    const rank = i + 1;
    const title = post.title.replace(/[#*_`]/g, '');
    const summary = (post.selftext || '').substring(0, 200).replace(/\n/g, ' ').trim();
    const summary2 = summary.length > 180 ? summary.substring(0, 180) + '...' : summary;

    lines.push(`### ${rank}. [${title}](https://www.reddit.com${post.permalink})`);
    lines.push(`Subreddit: r/${post.subreddit} | Score: ${post.score} | Comments: ${post.num_comments} | Upvote ratio: ${Math.round(post.upvote_ratio * 100)}%`);
    lines.push(`Posted: ~${post.hours_ago} hours ago`);
    lines.push('');
    lines.push(`**Summary:** ${summary2}`);
    lines.push('');
    lines.push(`**Viral Score:** ${post.viral_score.toFixed(1)}/10`);
    lines.push(`- Raw score: ${post.raw_points.toFixed(1)}/10`);
    lines.push(`- Engagement: ${post.engagement_points.toFixed(1)}/10`);
    lines.push(`- Upvote ratio: ${post.ratio_points.toFixed(1)}/10`);
    lines.push(`- Relevance bonus: ${post.relevance_bonus}/3`);
    lines.push('');

    if (post.mediaSources && post.mediaSources.length > 0) {
      lines.push('**Media:**');
      for (const src of post.mediaSources) {
        lines.push(`![](${src})`);
      }
      lines.push('');
    }
  }

  lines.push('## Honorable Mentions');
  lines.push('');
  for (let i = 5; i < Math.min(10, topPosts.length); i++) {
    const post = topPosts[i];
    const title = post.title.replace(/[#*_`]/g, '');
    const summary = (post.selftext || '').substring(0, 100).replace(/\n/g, ' ').trim();
    const summary2 = summary.length > 90 ? summary.substring(0, 90) + '...' : summary;
    lines.push(`### ${i + 1}. [${title}](https://www.reddit.com${post.permalink}) (r/${post.subreddit} | ${post.score} upvotes) – ${summary2}.`);
  }

  lines.push('');
  lines.push('## Media Summary');
  lines.push(`Downloaded images (${today}-media/):`);

  if (fs.existsSync(mediaDir)) {
    const files = fs.readdirSync(mediaDir).filter(f => !f.startsWith('.'));
    for (const file of files) {
      const stats = fs.statSync(path.join(mediaDir, file));
      const sizeKB = Math.round(stats.size / 1024);
      lines.push(`- **${file}** (${sizeKB} KB)`);
      lines.push(`  ![](${today}-media/${file})`);
    }
  } else {
    lines.push('No images downloaded.');
  }

  lines.push('');
  lines.push('---');
  lines.push(`**View on GitHub:** https://github.com/ozlemsultan90-cmyk/reddit-scout-reports/blob/main/reports/${today}.md`);

  return lines.join('\n');
}

async function run() {
  try {
    const topPosts = await processPosts();
    const report = formatReport(topPosts);
    fs.writeFileSync(reportFile, report);
    console.log(`Report generated: ${reportFile}`);
    return report;
  } catch (error) {
    console.error('Error generating report:', error);
    throw error;
  }
}

run();
